gemrb/core/Game.cpp

```diff
@@ -1737,44 +1756,100 @@ void Game::CastOnRest()
 	ieDword tmp = 0;
 	core->GetDictionary()->Lookup("Heal Party on Rest", tmp);
 	int specialCount = core->GetSpecialSpellsCount();
-
+	std::vector<Injuree> wholeparty;
 	if (tmp && specialCount != -1) {
 		int ps = GetPartySize(true);
 		int ps2 = ps;
-		std::vector<Actor *> injurees;
 		for (int idx = 1; idx <= ps; idx++) {
 			Actor *tar = FindPC(idx);
-			if (tar && tar->GetStat(IE_HITPOINTS) < tar->GetStat(IE_MAXHITPOINTS)) {
-				injurees.push_back(tar);
+			ieWord hpneeded=tar->GetStat(IE_MAXHITPOINTS) - tar->GetStat(IE_HITPOINTS);
+			if (tar && hpneeded > 0) {
+				wholeparty.push_back(Injuree(hpneeded,tar));
+			} else {
+				wholeparty.push_back(Injuree(0,tar));
 			}
-		}
-		if (injurees.empty()) injurees = PCs; // enable some nonhealing magic too
-
+			
+		}
+		// Following algorithm works thus:
+		// - If at any point there are no more injured party members, stop
+		// (amount of healing done is an estimation)
+		// - cast party members' all heal-all spells
+		// - repeat:
+		//       cast the most potent healing spell on the most injured member
 		SpecialSpellType *special_spells = core->GetSpecialSpells();
-		while (specialCount--) {
-			if (special_spells[specialCount].flags & SP_REST) {
-				if (injurees.empty()) injurees = PCs; // enable some nonhealing magic too
-				// check each party member for the spell
-				while (ps--) {
-					Actor *tar = GetPC(ps, true);
-					while (tar && tar->spellbook.HaveSpell(special_spells[specialCount].resref, 0)) {
-						// we have the spell, so cast it on most injured
-						// NOTE: no distinction is made about the potency of spells
-						Actor *injuree = injurees[0];
-						tar->DirectlyCastSpell(injuree, special_spells[specialCount].resref, 0, 1, true);
-						if (injuree->GetStat(IE_HITPOINTS) == injuree->GetStat(IE_MAXHITPOINTS)) {
-							if (!injurees.empty()) {
-								injurees.erase(injurees.begin());
+		if (specialCount != -1 ) {
+			std::sort(wholeparty.begin(),wholeparty.end());
+			specialCount = core->GetSpecialSpellsCount();
```

Comment:
this line looks redundant. you just checked specialCount so it had better be set to the correct value already.